### PR TITLE
[Doc] Ascend: refine T.tile.topk docstring, add test cases and API documentation

### PR DESCRIPTION
**Pull request by fredrekelthen**
## 1. 文件改动

| 文件 | 类型 | 改动说明 |
|------|------|----------|
| `docs/api_docs/T.tile.topk.md` | 新增 | 重写为校验完成版英文文档：删除 A5 行、修正 dst 大小约束（`aligned_topk`）、明确 repeatTimes 由 buffer 大小（非 actual_num）推导、补充 src 原地修改 dtype 差异、2D sum-of-dims 语义、32 元素对齐要求、输出格式（float32/float16 差异）、float16 index 精度上限、NaN/inf 行为 |
| `tilelang/language/ascend_tile.py` | 修改 | 精炼 `topk` docstring：修正 dst 大小要求（`aligned_topk` 而非 `2*K`）、补充 src 原地修改说明（float32 修改 / float16 不修改）、补充降序与输出格式说明|
| `testing/python/language/test_tilelang_ascend_language_topk.py` | 新增 | 完整测试套件：dtype 参数化（float16/float32）、actual_num 对齐/非对齐、K 边界（K=1/K=actual_num）、2D sum-of-dims 语义、src 原地修改验证（ascendc 严格 / pto 放宽）、大规模（4096）、float16 index 精度（2049/4096） |

---

## 2. 动态验证中发现的问题（真机 dav_m200/A2）

| # | 问题 | 验证结果 | 文档处理 |
|---|------|---------|---------|
| 1 | **dst 大小约束错误**：原文档写"至少 2×K 个元素"。实际 C++ `common.h:1667-1670` 中 `DataCopy(dst, sortDst, alignedTopk)`，其中 `alignedTopk = ceil(2K/(32/sizeof(T))) × (32/sizeof(T))`（float16 每块 16 元素、float32 每块 8 元素），拷贝量超过 2K | 编译+运行验证：K=10 float16 时 DataCopy 写 32 个元素（> 2K=20） | 约束 2 改为 `aligned_topk`；示例代码 dst 改为 32 元素 |
| 2 | **repeatTimes 推导错误**：原文档写 `repeatTimes = (actual_num + 31) // 32`。实际 `ascend_tile.py:451` 为 `repeatTimes = (max_actual_num + 31) // 32`，其中 `max_actual_num = sum(src.shape)`（src buffer 静态大小），与 actual_num 无关 | 生成代码确认：buffer (1,64) → repeatTimes=3（=ceil(65/32)），K=10 时 TopK 调用为 `(dst, src, tmp, 10, 3, 51)` | 约束 3/7 改为基于 buffer 大小 |
| 3 | **repeatTimes 范围 [0,255] 有误**：Sort32 的 `ASCENDC_CHECK_VALUE_RANGE(repeatTime, 0, 255)`，但 topk 的 repeatTimes 由 buffer 大小推导（≥1），实际范围 [1,255]，buffer 大小 ≤ 8160 | - | 约束 7 改为 [1, 255] |
| 4 | **非 32 倍数 buffer 索引损坏**：buffer_size < alignedCount 时（如 12 元素 buffer → alignedCount=32），-inf padding 越界写入邻接 tmp 区（索引区），导致 float32/float16 索引输出为 -inf | 验证：buffer_size=12 时输出索引全部 -inf；buffer_size=32（对齐）时索引正确 | 新增约束 4：src buffer 大小必须为 32 的倍数 |
| 5 | **src 原地修改未说明**：float32 路径在 src 上原地填充 -inf（`common.h:1504-1531`），float16 路径 cast 到 float32 后在 tmp 中填充，src 不被修改 | 验证：float32 尾部 [51:128) 变为 -inf（changed=77）；float16 unchanged（changed=0） | 2.2 参数表 src 标"输入/输出"；约束 9 区分 dtype |
| 6 | **2D 语义与 sort 不同**：topk 的 `max_actual_num = sum(dims)`（对 (M,N) 为 M+N，非乘积），仅前 M+N 个元素参与排序；与 sort 的"整体扁平化"不同 | 验证：(2,16) 且 actual_num=18（=sum）时结果正确；实际参与元素数 = sum(shape) | 2.3.2 说明 sum-of-dims 语义，推荐 1D 用法 |
| 7 | **A5 行需删除** | - | 按 FACTCHECK_WORKFLOW §2.2 删除 DataType 表 A5 行 |
| 8 | **K > actual_num**：硬件不报错，但超出 actual_num 的输出对无意义（padding -inf） | 验证：K=20 > actual_num=10 时编译运行成功，前 10 对正确 | 约束 6 明确 `1 <= K <= actual_num` |
| 9 | **NaN/inf 行为验证** | 验证：float32 输入 [nan,1,inf,2,3,nan,0,5] top-4 = [nan,nan,inf,5] | 约束 12 保留（nan 最先，inf 其次），与实测一致 |
| 10 | **dst==src 重叠** | 验证：dst==src 时结果错误（val ok=False） | 约束 10 保留 |
| 11 | **pto 后端 src padding 块粒度**：pto 路径 `ptoSort` 的 TFILLPAD_INPLACE 仅保证块对齐位置填充，部分块保留原值（`src/tl_templates/pto/common.h:1144-1151`） | 验证：pto float32 尾部 [51:64) min=-inf/max=0.0，[64:128) 全 -inf；top-K 结果仍正确 | 测试 src_inplace 对 pto 放宽断言；文档描述 ascendc 行为 |
| 12 | **float16 index 精度 >2048**：与 sort 相同机制（float32 排序后 CAST_RINT 回 half），index 超过 2048 精度损失 | 验证：2049/4096 时索引 mismatch 但数值正确 | 2.3.5 注明 |

---

## 3. 测试结果

| 测试函数 | 用例数 | 测什么 | 结果 |
|----------|:---:|--------|:----:|
| `test_topk_basic` | 12 | dtype(2) × target(2) × actual_num(128/51/100) | 全部 PASSED |
| `test_topk_k_boundary` | 8 | dtype(2) × target(2) × K(1/64) | 全部 PASSED |
| `test_topk_2d` | 4 | dtype(2) × target(2)，(2,16) buffer + actual_num=18 | 全部 PASSED |
| `test_topk_src_inplace` | 4 | dtype(2) × target(2)，src 原地修改验证 | 全部 PASSED |
| `test_topk_large_actual_num` | 2 | float32 × target(2)，actual_num=4096 | 全部 PASSED |
| `test_topk_fp16_index_precision` | 4 | float16 × target(2) × actual_num(2049/4096) | 全部 PASSED |

- 默认用例（ascendc，非 low_priority）：**14 passed**（真机 dav_m200/A2）
- low_priority 用例（pto + 大规模 + fp16 精度）：**20 passed**（分批运行，避免 aicore 异常累积）

---

## 4. pytest 标签

| 测试函数 | low_priority | PR 执行 |
|----------|:---:|:---:|
| `test_topk_basic` pto 用例 | 6 | 跳过 |
| `test_topk_k_boundary` pto 用例 | 4 | 跳过 |
| `test_topk_2d` pto 用例 | 2 | 跳过 |
| `test_topk_src_inplace` pto 用例 | 2 | 跳过 |
| `test_topk_large_actual_num` 全部 | 2 | 跳过 |
| `test_topk_fp16_index_precision` 全部 | 4 | 跳过 |
| ascendc 默认用例 | — | 14 执行 |

---

## 5. 验证

- 语法检查：`ast.parse` 通过（ascend_tile.py、测试文件）
- 测试收集：34 个用例正常收集
- 默认测试（`not (low_priority or ci_skip)`）：**14 passed**（真机 dav_m200/A2）
- 文档约束与源码 `common.h` / `ascend_tile.py` / 真机行为逐一核对一致

### CODE REVIEW

**Review by LLMZhangYC — DISMISSED** [review]
lgtm

File: docs/api_docs/T.tile.topk.md [hunk 1]
```diff
@@ -0,0 +1,108 @@
+# T.tile.topk
+
+## 1. Description
+
+Sorts the source data in descending order and extracts the top K interleaved (value, index) pairs: `dst = [val0, idx0, val1, idx1, ..., val(K-1), idx(K-1)]`, where `idx` is the 0-based position in the aligned buffer before sorting (including -inf padding positions). Internally, the full sort is performed first (via sort32 + merge), then the top K pairs are copied to `dst`.
```
Comment (2Sheldonlaw) [inline docs/api_docs/T.tile.topk.md:5]: 文档改成中文描述
同时测试过程中出现错误信息不需要描述在文档里面，如果不支持先限制住

Comment (fredrekelthen) [reply docs/api_docs/T.tile.topk.md]: 已在 commit babb0123 中将文档改为中文，并移除了测试过程中错误信息的描述，不支持的场景已在约束中限制说明。

File: tilelang/language/ascend_tile.py [hunk 2]
```diff
@@ -557,20 +557,27 @@ def topk(
...
+             aligned_topk = ((2*K + elems_per_block - 1) / elems_per_block) *
+             elems_per_block and elems_per_block = 32 / sizeof(dtype).
         src: Source buffer containing the data to find top K from.
-             Assumes src has static shape for buffer sizing.
-        K: Number of top elements to extract.
-        actual_num: The number of valid elements in src (can be symbolic for dynamic shapes).
+             Must have a compile-time static shape. For float32, positions from
+             actual_num to aligned_count are padded with -inf in-place; for
+             float16, src is not modified (internally cast to float32).
+        K: Number of top elements to extract, 1 <= K <= actual_num.
```
Comment (LLMZhangYC) [inline tilelang/language/ascend_tile.py:575]: docstring 声明 `1 <= K <= actual_num`，但实现侧没有看到对应的参数校验。K 为 0、负数或大于 actual_num 时，后续对 top-k 结果的切片/拷贝可能产生越界访问或错误结果。建议在构造 intrinsic 前显式检查 K 和 actual_num，并抛出 ValueError，同时为这些非法输入补充测试。

File: docs/api_docs/T.tile.topk.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    *,
+    tmp: Buffer | BufferRegion | None = None,
+)
+```
+
+### 2.2 参数
+
+| 参数 | 方向 | 说明 | 类型 | 必须/可选 |
+|------|------|------|------|-----------|
+| dst | 输出 | 存储前 K 个交错的（值, 索引）对。必须至少包含 `aligned_topk` 个元素（`aligned_topk = ((2*K + elems_per_block - 1) / elems_per_block) * elems_per_block`，其中 `elems_per_block = 32 / sizeof(T)`，即 float16 为 16，float32 为 8） | tensor | 必须 |
```
Comment (LLMZhangYC) [inline docs/api_docs/T.tile.topk.md:26]: 文档要求 dst 至少包含 aligned_topk 个元素，但没有说明容量不足时是否会被 API 主动拒绝。若当前实现只将 dst 指针传给底层 intrinsic，容量不足可能导致静默越界写入。建议增加 extent 校验，并补充 dst 容量不足的负例测试。

File: testing/python/language/test_tilelang_ascend_language_topk.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+import pytest
+import torch
+
+import tilelang
+import tilelang.language as T
+
+"""
+Test suite for T.tile.topk API.
+
+Covers:
```
Comment (LLMZhangYC) [inline testing/python/language/test_tilelang_ascend_language_topk.py:12]: 当前测试覆盖了常规 K 值和边界值，但没有覆盖 K=0、K>actual_num、actual_num=0、src 容量不足以及 dst 仅有 2*K 而不足 aligned_topk 的情况。建议增加这些非法参数测试，确保文档中声明的约束在实现中得到验证，而不是依赖底层硬件异常。

File: tilelang/language/ascend_tile.py (hunk 2 above)
Comment (fredrekelthen) [reply tilelang/language/ascend_tile.py:575]: 已采纳：在 topk 函数中增加了编译期参数校验——K < 1、actual_num < 1 或 K > actual_num（均为 tir.IntImm 时）抛 ValueError。对于动态值，无前端校验，由硬件处理。

File: docs/api_docs/T.tile.topk.md (hunk 3 above)
Comment (fredrekelthen) [reply docs/api_docs/T.tile.topk.md:26]: 已采纳：文档约束 2 补充说明 dst/src 容量不足时无编译期校验，可能产生静默越界写入，调用方需自行确保容量。编译期 extent 校验可后续在 topk 实现中增强，当前文档 PR 先行标注风险。

File: testing/python/language/test_tilelang_ascend_language_topk.py (hunk 4 above)
Comment (fredrekelthen) [reply testing/python/language/test_tilelang_ascend_language_topk.py:12]: 已采纳建议。增加了编译期校验后，可覆盖 K=0/K<1、actual_num=0、K>actual_num 的负例（编译期抛 ValueError）。dst 容量不足的测试尚无法在编译期捕获（未增加 extent 校验），已在文档约束 2 标注风险。后续可在实现中增强 extent 校验后补充对应测试。

**Review by ChaoyangJi — APPROVED** [review]
approve

**Review by LLMZhangYC — APPROVED** [review]
approve
